Optimize tadir read

fmabap opened the pull request:
Closes #6428

larshp commented:
guess the expected TADIR entries doesnt exist in open-abap

anyhow, suggest removing the assumptions, move the test to `/test/`, and use osql test doubles for TADIR, example in https://github.com/open-abap/open-abap-core/blob/main/src/unit/osql/cl_osql_test_environment.clas.testclasses.abap

larshp commented on test/zcl_abapgit_tadir_test.clas.testclasses.abap:
```diff
@@ -0,0 +1,169 @@
+CLASS ltcl_build DEFINITION DEFERRED.
+
+CLASS ltcl_build DEFINITION FOR TESTING RISK LEVEL HARMLESS DURATION SHORT FINAL.
```
its not deferred, its defined in the next statement, MEDIUM runtime will make it execute later than the tests in /src/

```suggestion
CLASS ltcl_build DEFINITION FOR TESTING RISK LEVEL HARMLESS DURATION MEDIUM FINAL.
```

fmabap replied:
I adjusted it, but now the unit test failed and I don't know why. In Eclipse in the src folder was it ok.
File \test\zcl_abapgit_tadir_test.clas.testclasses.abap

fmabap replied:
<img width="858" alt="image" src="https://github.com/abapGit/abapGit/assets/101019006/93703b64-e365-49ee-9f01-28a5011095af">

larshp replied:
fixed with https://github.com/abapGit/abapGit/pull/6429#issuecomment-1691678003

larshp commented:
https://github.com/open-abap/open-abap-core/pull/761

fmabap commented on src/zif_abapgit_definitions.intf.abap:
```diff
@@ -3,6 +3,13 @@ INTERFACE zif_abapgit_definitions
+      obj_name TYPE tadir-obj_name,
+    END OF ty_obj.
+  TYPES:
+    ty_obj_tt TYPE STANDARD TABLE OF ty_obj WITH DEFAULT KEY.
```
I changed it from binary search to sorted table with non-unique key obj_type and hased unique key obj_type_obj_name

larshp commented:
keep the scope of this PR to optimize the TADIR read? as per https://github.com/abapGit/abapGit/issues/6428

larshp commented:
https://github.com/abaplint/transpiler/pull/1304

fmabap commented:
The only scope of this PR is to optimize the TADIR read as per https://github.com/abapGit/abapGit/issues/6428

fmabap commented:
> > The only scope of this PR is to optimize the TADIR read as per #6428
> 
> then I dont expect any changes to `src/objects/core/zcl_abapgit_serialize.clas.abap`, https://github.com/abapGit/abapGit/pull/6429/files#diff-e79d9e936a1a8f68ab662c89a2374f7e84b86b8b3686e3677971ad362ae23ef9

The changes in this class zcl_abapgit_serialize are required to pass the filter to the TADIR class.

larshp commented:
how can the performance problem be reproduced in the current codebase? And if so, which parts are part of that performance optimization, vs which parts is the new unused functionality being introduced?

fmabap commented:
> how can the performance problem be reproduced in the current codebase? And if so, which parts are part of that performance optimization, vs which parts is the new unused functionality being introduced?

You would need a big repository to reproduce directly in the system. I created the demo branch https://github.com/fmabap/abapgit/tree/demo_%236428_issue to show the issue with a unit test on the SAP Standard top package BASIS with 400.000 objects and multiple namespaces.
<img width="1491" alt="image" src="https://github.com/abapGit/abapGit/assets/101019006/59f9f26c-4283-4ba8-a403-d8be3e4890ce">

The main issue was the check if the namespace was alread determined. There was a check for each object with a namespace in the table with the 400.000 entries if the namespace exists, else it would be added.

<img width="496" alt="image" src="https://github.com/abapGit/abapGit/assets/101019006/cc3dff44-cab5-497d-a1a1-ae7dc15e2baf">

I solved it be adding an additional table that contains only the namespaces. There should be not more than 100 namespaces in real live.

<img width="662" alt="image" src="https://github.com/abapGit/abapGit/assets/101019006/b9cd6df5-304c-409c-9fcd-bc2f449514f8">

<img width="496" alt="image" src="https://github.com/abapGit/abapGit/assets/101019006/42cd54a6-2000-454d-8e36-c24be1d433d6">

We use the filter functionality daily in our project with the Stage by Transport functionality.

<img width="358" alt="image" src="https://github.com/abapGit/abapGit/assets/101019006/50b74109-d7c4-4ef8-be55-c85b712b0c3f">

Passing the filter to the tadir class avoids the CHECK_EXISTS for the not relevant objects. 

<img width="466" alt="image" src="https://github.com/abapGit/abapGit/assets/101019006/95023a43-5c5c-4aed-8ddc-7ab3ae794b11">

fmabap commented:
@larshp  I added more infos to the fail of the unit tests.
<img width="1033" alt="image" src="https://github.com/abapGit/abapGit/assets/101019006/8edcb614-4af3-4bda-8311-ec4fc5ebfd5f">

As mentioned it works fine in Eclipse
<img width="370" alt="image" src="https://github.com/abapGit/abapGit/assets/101019006/7e60947f-136e-478a-b357-67e679d4cc5e">

It looks like that the transpiler has an issue with these read tables
<img width="521" alt="image" src="https://github.com/abapGit/abapGit/assets/101019006/6e95a90d-eeea-4256-84f0-fd83fa6450bd">

May be it is because of the special type of the it_filter table that @mbtools requested.

<img width="509" alt="image" src="https://github.com/abapGit/abapGit/assets/101019006/ecd06c76-b39f-4c84-8789-5b61799c196b">

larshp commented on src/objects/sap/zif_abapgit_sap_namespace.intf.abap:
```diff
@@ -3,14 +3,22 @@ INTERFACE zif_abapgit_sap_namespace
+  METHODS split_by_name
+    IMPORTING
+      iv_obj_with_namespace    TYPE tadir-obj_name
+    EXPORTING
+      ev_namespace             TYPE trnspace-namespace
+      ev_obj_without_namespace TYPE tadir-obj_name
```
returning a structure will make it possible to do chaining in 702

fmabap replied:
I adjusted it.

larshp commented:
suggest moving the refactoring of calls to `RS_NAME_SPLIT_NAMESPACE` to a separate PR

fmabap commented:
> suggest moving the refactoring of calls to `RS_NAME_SPLIT_NAMESPACE` to a separate PR

I created the PR #6443 for it.

fmabap commented:
@larshp the unit test for the namespace filter works after your fix, but the unit test for the filter of the local package $SWF_RUN_CNT still fails on github, but in eclipse it is fine.

![image](https://github.com/abapGit/abapGit/assets/101019006/ffe02d0b-8de9-4fe0-966c-a8cf6b1d49ed)

The packages are selected here 

![image](https://github.com/abapGit/abapGit/assets/101019006/2ebeaa69-749a-4108-9a9b-42bdfa5ec3cd)

The TADIR entry for them are created here

![image](https://github.com/abapGit/abapGit/assets/101019006/44de7195-a634-47be-8d34-cbb698479332)

Can you please check, why this unit test does not work on github?

fmabap commented:
@larshp 
I added additional unit tests to find the issue of the abap transpiler.

It looks like that issue is this while loop.

![image](https://github.com/abapGit/abapGit/assets/101019006/ecea5ef7-0049-46e2-b467-5b1dc07b50a2)

I added unit tests that did the selection without the while they work also on github

![image](https://github.com/abapGit/abapGit/assets/101019006/2f59f56f-1dcf-479a-9ec5-1f00e0f08f87)

Can you please fix it?

fmabap commented:
@larshp 
The PR #6445 shows the transpile error in the new test class abapGit\test\zcl_abapgit_sap_package_test.clas.testclasses.abap without the new logic.

fmabap commented:
All checks passed 🎉🎉🎉 => ready for review

fmabap commented:
This PR has been splitted into #6461 and #6462
